File: Array/2.Medium/01.2_sum_problem.cpp
/*
QUESITON:-
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
You may assume that each input would have exactly one solution, and you may not use the same element twice.
You can return the answer in any order.

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
*/


#include <bits/stdc++.h>
using namespace std;
/*
Approach:

-> Create an empty map to store the elements and their corresponding indices.
>  Iterate through the input array, nums, and for each element:
    Calculate the complement by subtracting the current element from the target value.
    Check if the complement exists in the map.
    If the complement exists, return the indices of the current element and the complement.
    If the complement does not exist, add the current element and its index to the map.
-> If no solution is found, return an empty vector or a message indicating no solution exists.
*/

// CODE:-
vector<int> twoSum(vector<int> &nums, int target)
{
    unordered_map<int, int> mp;
    for (int i = 0; i < nums.size(); i++)
    {
        int remain = target - nums[i];
        if (mp.find(remain) != mp.end() && mp[remain] != i)
            return {i, mp[remain]};
        mp[nums[i]] = i;
    }
    return {-1, -1};
    // If the question asks to just return whether pair exists or not, not the indexes in that case we can sort and easily find the pair sum without extra space
}


int main()
{
    // Test case 1
    vector<int> nums1 = {2, 7, 11, 15};
    int target1 = 9;
    vector<int> result1 = twoSum(nums1, target1);
    cout << "Test 1: nums = [2,7,11,15], target = 9" << endl;
    cout << "Output: [" << result1[0] << "," << result1[1] << "]" << endl;
    
    // Test case 2
    vector<int> nums2 = {3, 2, 4};
    int target2 = 6;
    vector<int> result2 = twoSum(nums2, target2);
    cout << "\nTest 2: nums = [3,2,4], target = 6" << endl;
    cout << "Output: [" << result2[0] << "," << result2[1] << "]" << endl;
    
    return 0;
}

// TIME COMPLEXITY = O(N)

File: 01.Arrays/2.Medium/01.2_sum_problem.py
# QUESTION:-
# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
# You may assume that each input would have exactly one solution, and you may not use the same element twice.
# Example 1: Input: nums = [2,7,11,15], target = 9, Output: [0,1]
# Example 2: Input: nums = [3,2,4], target = 6, Output: [1,2]

# APPROACH:-
# Create a dictionary to store elements and their indices
# For each element, calculate complement = target - element
# If complement exists in dict, return indices
# Otherwise, add current element and index to dict

# CODE:-
def twoSum(nums, target):
    """Find two numbers that add up to target"""
    mp = {}  # dictionary to store element -> index
    for i in range(len(nums)):
        remain = target - nums[i]
        if remain in mp:
            return [i, mp[remain]]
        mp[nums[i]] = i
    return [-1, -1]  # No solution found


if __name__ == '__main__':
    # Test case 1
    print("Test 1:")
    nums1 = [2, 7, 11, 15]
    target1 = 9
    result1 = twoSum(nums1, target1)
    print(f"Output: {result1}")
    print("Expected: [0, 1]")
    print()
    
    # Test case 2
    print("Test 2:")
    nums2 = [3, 2, 4]
    target2 = 6
    result2 = twoSum(nums2, target2)
    print(f"Output: {result2}")
    print("Expected: [1, 2]")

# TIME COMPLEXITY = O(N)
# SPACE COMPLEXITY = O(N) (for the hashmap)
